angel_system/impls/detect_activities/swinb/swinb_detect_activities.py

```diff
@@ -0,0 +1,211 @@
+        weights for the model.
+    :param num_classes: Number of classes the model was trained on.
+    :param labels_file: Path to the labels file for the given checkpoint.
+        The labels file is a text file with the class labels, one class
+        per line.
+    :param num_frames: Number of frames passed to the model for inference.
+    :param sampling_rate: Sampling rate for the frame input. For example,
+        if this is set to 2 and num_frames is set to 32, the activity
+        detector should pass 64 frames as input to the detect activities
+        function.
+    :param use_cuda: Attempt to use a cuda device for inferences. If no
+        device is found, CPU is used.
+    :param cuda_device: When using CUDA use the device by the given ID. By
+        default, this refers to GPU ID 0. This parameter is not used if
+        `use_cuda` is false.
+    :param det_threshold: Threshold for which predictions must exceed to
+        create an activity detection.
+    """
+
+    def __init__(
+        self,
+        checkpoint_path: str,
+        num_classes: int,
+        labels_file: str,
+        num_frames: int = 32,
+        sampling_rate: int = 2,
+        use_cuda: bool = False,
+        cuda_device: Union[int, str] = "cuda:0",
+        det_threshold: float = 0.75,
+    ):
+        self._checkpoint_path = checkpoint_path
+        self._num_classes = num_classes
+        self._use_cuda = use_cuda
+        self._cuda_device = cuda_device
+        self._det_threshold = det_threshold
+        self._num_frames = num_frames
+        self._sampling_rate = sampling_rate
+        self._labels_file = labels_file
+
+        # Set to None for lazy loading later.
+        self._model: torch.nn.Module = None  # type: ignore
+        self._model_device: torch.device = None  # type: ignore
+
+        # Default configs from learn SwinVideo config
+        self._mean = [0.45, 0.45, 0.45]
+        self._std = [0.225, 0.225, 0.225]
+        self._crop_size = 224
+        self._frames_per_second = 30
+
+        # Transfrom from learn/TimeSformer/video_classification.py
+        self.transform = torchvision.transforms.Compose(
+            [
+                torchvision.transforms.ToTensor(),
+                torchvision.transforms.Normalize(
+                    self._mean, self._std
+                )
+            ]
+        )
+
+        # Set up the labels from the given labels file
+        self._labels = []
+        with open(self._labels_file, "r") as f:
+            for line in f:
+                self._labels.append(line.rstrip())
+
+    def get_model(self) -> "torch.nn.Module":
+        """
+        Lazy load the torch model in an idempotent manner.
+        :raises RuntimeError: Use of CUDA was requested but is not available.
+        """
+        model = self._model
+        if model is None:
+            # Load the model with the checkpoint
+            model = swin_b(self._checkpoint_path, self._num_classes)
```

Comment:
This might not matter for the near term, but sometimes we are given models that are already encoded to "load" onto a GPU, and have in previous implementations used a load-call that had a parameter to load to a specific given device instead of what was encoded in the weights serialization. Is that tenable here?

Reply:
Hmm I'm not sure if I understand. Is that like using `torch.load` with the `map_location` parameter?

Reply:
Yes, exactly.